### PR TITLE
[IMP] payment(_paypal): UX and payment flow improvements

### PR DESCRIPTION
**Pull request by anko-odoo**
UX was lacking comparing to other payment providers,
important fields were not always shown or were checkboxes
when they should be automatically true.

In order to make payment flow easier and more intitive, unnecessary fields
were removed, email is automatically filled. Now when user cancels transaction
on paypal before paying, it automatically cancels transaction on Odoo. Additionaly, quick
onboarding is only available if user already has paypal account and Stripe
no longer installs ond configures paypal if Stripe's onboarding get canceled.

task-2854184

See also:
- https://github.com/odoo/upgrade/pull/4025
- https://github.com/odoo/internal/pull/2076
- https://github.com/odoo/documentation/pull/3063

### CODE REVIEW

File: addons/payment_paypal/controllers/main.py [hunk 1]
```diff
@@ -46,23 +47,30 @@ def paypal_return_from_checkout(self, **pdt_data):
...
 
         return request.redirect('/payment/status')
 
+    @http.route(
+        _cancel_url, type='http', auth='public', methods=['GET', 'POST'], csrf=False,
+        save_session=False
+    )
+    def paypal_return_from_checkout_before_paying(self, transaction_reference=None, **pdt_data):
+        # The customer has canceled payment
+        _logger.info("handling redirection for cancalation PayPal with data:\n%s", pprint.pformat(pdt_data))
```
Comment (PermanAtayev) [inline addons/payment_paypal/controllers/main.py:70]:
```suggestion
        _logger.info("handling redirection for cancellation PayPal with data:\n%s", pprint.pformat(pdt_data))
```

**Review by PermanAtayev — COMMENTED** [review]
Nitpick :)

**Review by chevalierv — CHANGES_REQUESTED** [review]
Good job overall :+1:

File: addons/payment_paypal/controllers/main.py [hunk 2]
```diff
@@ -46,23 +47,30 @@ def paypal_return_from_checkout(self, **pdt_data):
...
+    @http.route(
+        _cancel_url, type='http', auth='public', methods=['GET', 'POST'], csrf=False,
+        save_session=False
+    )
+    def paypal_return_from_checkout_before_paying(self, transaction_reference=None, **pdt_data):
+        # The customer has canceled payment
+        _logger.info("handling redirection for cancallation PayPal with data:\n%s", pprint.pformat(pdt_data))
+        request.env['payment.transaction'].sudo()._cancel_tx(transaction_reference)
+        return request.redirect('/payment/status')
+
```
[the remark this answers is not in the document]
Comment (anko-odoo) [reply addons/payment_paypal/controllers/main.py:73]: I thought it would be cleared since I want to pass  f'?transaction_reference={self.reference}' in url.
The third comment I don't really understand, so please elaborate, but from what I understand I have a question how would we cancel transaction if we wouldn't know which one?

[the remark this answers is not in the document]
Comment (chevalierv) [reply addons/payment_paypal/controllers/main.py:73]: Sorry for the confusion. 

What I meant is that creating a method `_cancel_tx` in another file just for two lines seems overkill. I'm pretty sure that you can use `tx_sudo._handle_notification_data('paypal', notification_data)`, which will call the override of `_process_notification_data` in `payment_payment/model/payment_transaction.py`. By doing that, we stay in the standard flow and ensure that the behavior will stay the same for all flows. 

Finally, I think that we could reuse the return_url but I'm fine if you want to create a new route as long as you call `_handle_notification_data` in the end.

I hope that it's clearer now, but we can discuss it orally if you still have questions :)

File: addons/payment_paypal/controllers/main.py [hunk 3]
```diff
@@ -16,6 +16,7 @@
 
 class PaypalController(http.Controller):
     _return_url = '/payment/paypal/return/'
+    _cancel_url = '/payment/paypal/cancel'
```
Comment (chevalierv) [inline addons/payment_paypal/controllers/main.py:19]:
```suggestion
    _cancel_url = '/payment/paypal/cancel/'
```

File: addons/payment_paypal/models/payment_provider.py [hunk 4]
```diff
@@ -17,9 +17,8 @@ class PaymentProvider(models.Model):
     paypal_email_account = fields.Char(
         string="Email",
         help="The public business email solely used to identify the account with PayPal",
-        required_if_provider='paypal')
-    paypal_seller_account = fields.Char(
-        string="Merchant Account ID", groups='base.group_system')
+        required_if_provider='paypal',
+        default=lambda self: self.env.company.email)
     paypal_pdt_token = fields.Char(string="PDT Identity Token", groups='base.group_system')
     paypal_use_ipn = fields.Boolean(
```
[the remark this answers is not in the document]
Comment (anko-odoo) [reply addons/payment_paypal/models/payment_provider.py]: Done :)

File: addons/payment_paypal/controllers/main.py (hunk 3 above)
Comment (anko-odoo) [reply addons/payment_paypal/controllers/main.py:19]: Done

File: addons/payment/wizards/payment_onboarding_views.xml [hunk 5]
```diff
@@ -16,7 +16,6 @@
                                 <group>
                                     <field name="paypal_user_type" widget="radio" nolabel="1" colspan="2"/>
                                     <field name="paypal_email_account" attrs="{'required': [('payment_method', '=', 'paypal')]}" string="Email"/>
-                                    <field name="paypal_seller_account" attrs="{'invisible': [('paypal_user_type', '=', 'new_user')], 'required': [('paypal_user_type', '!=', 'new_user'), ('payment_method', '=', 'paypal')]}" />
                                     <field name="paypal_pdt_token" password="True" attrs="{'invisible': [('paypal_user_type', '=', 'new_user')], 'required': [('paypal_user_type', '!=', 'new_user'), ('payment_method', '=', 'paypal')]}" />
```
[the remark this answers is not in the document]
Comment (anko-odoo) [reply addons/payment/wizards/payment_onboarding_views.xml]: Done :)

File: addons/payment_paypal/controllers/main.py [hunk 6]
```diff
@@ -46,20 +47,33 @@ def paypal_return_from_checkout(self, **pdt_data):
...
+    )
+    def paypal_return_from_checkout_before_paying(self, transaction_reference=None, **pdt_data):
+        # The customer has canceled payment
+        _logger.info("handling redirection for cancellation PayPal with data:\n%s", pprint.pformat(pdt_data))
+        tx_sudo = request.env['payment.transaction'].sudo().search([
+            ('reference', '=', transaction_reference),
+            ('provider_code', '=', 'paypal'),
+        ])
+        notification_data = self._verify_pdt_notification_origin(pdt_data, tx_sudo=tx_sudo)
+        tx_sudo._handle_notification_data('paypal', notification_data)
```
Comment (chevalierv) [inline addons/payment_paypal/controllers/main.py:76]: @Feyensv pointed out to me that we need to check the provenance of the request to avoid people canceling transactions just by making a request to this endpoint. 

`_verify_pdt_notification_origin` seemed to be a good idea except that Paypal doesn't send anything back to Odoo in that case so we can't verify anything. To check the origin of the request, you should use an access token that we can check in this controller. There is some method for that in `odoo/addons/payment/utils.py`. Have a look at `generate_access_token` and `check_access_token`. 

Since we need something to be sure that the request comes from this database, we'll use the `database_uuid` to generate the access token. That will guarantee that the request is not forged and comes from this database :)

Please find one of my suggestions below and the other in `addons/payment_paypal/models/payment_transaction.py`. 

```suggestion
        database_uuid = request.env['ir.config_parameter'].sudo().get_param('database.uuid')
        if not payment_utils.check_access_token(access_token, database_uuid, transaction_reference):
            raise Forbidden()
        tx_sudo._handle_notification_data('paypal', {})```

**Review by chevalierv — CHANGES_REQUESTED** [review]
Here is my second pass :duck: 

A bit of nitpicking and some things that I didn't catch the first time. 

Feel free to reach out if you need any help with my comments.

Btw, it could be nice to remove unused dependencies while we're making changes in those files ^^ I saw one unused dependency in `odoo/addons/payment_paypal/models/payment_transaction.py`, but I didn't check the other files.

File: addons/payment/wizards/payment_onboarding_wizard.py [hunk 7]
```diff
@@ -96,7 +90,6 @@ def add_payment_methods(self):
             if self.payment_method == 'paypal':
                 new_env.ref('payment.payment_provider_paypal').write({
                     'paypal_email_account': self.paypal_email_account,
-                    'paypal_seller_account': self.paypal_seller_account,
                     'paypal_pdt_token': self.paypal_pdt_token,
                     'state': 'enabled',
```
Comment (chevalierv) [inline addons/payment/wizards/payment_onboarding_wizard.py:94]: I think we forgot to publish the provider when the user is setting it up through the onboarding payment wizard. Could you make the change here and a few lines below (if the `payment_method` is "manual")?
```suggestion
                    'state': 'enabled',
                    'is_published': True
```

File: addons/payment_paypal/controllers/main.py [hunk 8]
```diff
@@ -46,20 +47,33 @@ def paypal_return_from_checkout(self, **pdt_data):
         retrieved and with it the transaction which will be immediately post-processed.
         """
         _logger.info("handling redirection from PayPal with data:\n%s", pprint.pformat(pdt_data))
-        if not pdt_data:  # The customer has canceled or paid then clicked on "Return to Merchant"
-            pass  # Redirect them to the status page to browse the (currently) draft transaction
-        else:
             # Check the origin of the notification
```
[the remark this answers is not in the document]
Comment (anko-odoo) [reply addons/payment_paypal/controllers/main.py:50]: Done.

File: addons/payment_paypal/controllers/main.py [hunk 9]
```diff
@@ -46,20 +47,33 @@ def paypal_return_from_checkout(self, **pdt_data):
...
-            else:
+        tx_sudo = request.env['payment.transaction'].sudo()._get_tx_from_notification_data(
+            'paypal', pdt_data
+        )
+        try:
+            notification_data = self._verify_pdt_notification_origin(pdt_data, tx_sudo)
+        except Forbidden:
+            _logger.exception("could not verify the origin of the PDT; discarding it")
+        else:
                 # Handle the notification data
```
[the remark this answers is not in the document]
Comment (anko-odoo) [reply addons/payment_paypal/controllers/main.py:59]: Done.

File: addons/payment_paypal/controllers/main.py [hunk 10]
```diff
@@ -16,6 +16,7 @@
 
 class PaypalController(http.Controller):
     _return_url = '/payment/paypal/return/'
+    _cancel_url = '/payment/paypal/cancel/'
     _webhook_url = '/payment/paypal/webhook/'
```
[the remark this answers is not in the document]
Comment (anko-odoo) [reply addons/payment_paypal/controllers/main.py:20]: Done

File: addons/payment_paypal/controllers/main.py [hunk 11]
```diff
@@ -46,20 +47,33 @@ def paypal_return_from_checkout(self, **pdt_data):
...
+            _logger.exception("could not verify the origin of the PDT; discarding it")
+        else:
                 # Handle the notification data
-                tx_sudo._handle_notification_data('paypal', notification_data)
+            tx_sudo._handle_notification_data('paypal', notification_data)
+
+        return request.redirect('/payment/status')
+
+    @http.route(
+        _cancel_url, type='http', auth='public', methods=['GET', 'POST'], csrf=False,
```
[the remark this answers is not in the document]
Comment (anko-odoo) [reply addons/payment_paypal/controllers/main.py:65]: Done

File: addons/payment_paypal/models/payment_transaction.py [hunk 12]
```diff
@@ -92,36 +93,35 @@ def _process_notification_data(self, notification_data):
...
-            self.provider_id._paypal_send_configuration_reminder()
-
-        if payment_status in PAYMENT_STATUS_MAPPING['pending']:
-            self._set_pending(state_message=notification_data.get('pending_reason'))
-        elif payment_status in PAYMENT_STATUS_MAPPING['done']:
-            self._set_done()
-        elif payment_status in PAYMENT_STATUS_MAPPING['cancel']:
+        if not notification_data:
             self._set_canceled()
+
```
[the remark this answers is not in the document]
Comment (anko-odoo) [reply addons/payment_paypal/models/payment_transaction.py:98]: Can you elaborate a bit about this one? Because since there is no notification_data, ValidationError from line 102 is raised right?

[the remark this answers is not in the document]
Comment (anko-odoo) [reply addons/payment_paypal/models/payment_transaction.py:98]: Yes, I realised a moment before your comment :D Thanks anyway for explanation.

File: addons/payment_paypal/controllers/main.py [hunk 13]
```diff
@@ -46,20 +47,33 @@ def paypal_return_from_checkout(self, **pdt_data):
...
+        save_session=False
+    )
+    def paypal_return_from_checkout_before_paying(self, transaction_reference=None, **pdt_data):
+        # The customer has canceled payment
+        _logger.info("handling redirection for cancellation PayPal with data:\n%s", pprint.pformat(pdt_data))
+        tx_sudo = request.env['payment.transaction'].sudo().search([
+            ('reference', '=', transaction_reference),
+            ('provider_code', '=', 'paypal'),
+        ])
```
[the remark this answers is not in the document]
Comment (anko-odoo) [reply addons/payment_paypal/controllers/main.py:74]: Could you elaborate a bit on this? Because in normal case we check if the reference we got from paypal exists in odoo, and in canceled case the reference is what we give so it always will be in there right?

[the remark this answers is not in the document]
Comment (chevalierv) [reply addons/payment_paypal/controllers/main.py:74]: You're right, and your code works and is good, but `_get_tx_from_notification_data` was created purposefully for that. This is a method from the payment engine that is overridden in every payment provider to find the transaction based on the data sent back by the provider. 

If you look closely at `_handle_notification_data`, you'll see that the first call is also to `_get_tx_from_notification_data`. Using `_get_tx_from_notification_data` seems then the right choice to ensure that the search domain will be the same. 

Basically, I invite you to always use existing methods to keep the responsibility centralized and avoid duplicating code.

File: addons/payment_paypal/controllers/main.py (hunk 6 above)
Comment (chevalierv) [reply addons/payment_paypal/controllers/main.py:76]: I updated my suggestion following our discussion on discord. There is no need to use the `database_uuid`. Sorry for the confusion

File: addons/payment_paypal/models/payment_transaction.py [hunk 14]
```diff
@@ -35,12 +35,22 @@ def _get_specific_rendering_values(self, processing_values):
             return res
 
         base_url = self.provider_id.get_base_url()
+        #TO DO It work with plus, does not work witkout.
+        #canceled_tx_details = f'?transaction_reference={self.reference}&access_token={payment_utils.generate_access_token(self.reference)}'
+        #canceled_tx_details = f'?{urls.url_encode({'transaction_reference':self.reference,'access_token':payment_utils.generate_access_token(self.reference)})}'
```
Comment (chevalierv) [inline addons/payment_paypal/models/payment_transaction.py:40]: Shall we keep these comments?

**Comment by chevalierv** [comment]
LGTM :+1:

**Review by chevalierv — APPROVED** [review]
LGTM :+1:

File: addons/payment_paypal/models/payment_transaction.py [hunk 15]
```diff
@@ -51,7 +57,7 @@ def _get_specific_rendering_values(self, processing_values):
             'item_number': self.reference,
             'last_name': partner_last_name,
             'lc': self.partner_lang,
-            'notify_url': webhook_url if self.provider_id.paypal_use_ipn else None,
+            'notify_url': webhook_url,
```
Comment (AntoineVDV) [inline addons/payment_paypal/models/payment_transaction.py]: This means that the `t-if` can be removed from the redirect form template.

File: addons/payment_paypal/tests/test_paypal.py [hunk 16]
```diff
@@ -1,4 +1,5 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
+from werkzeug import urls
```
Comment (AntoineVDV) [inline addons/payment_paypal/tests/test_paypal.py]: Sort imports in this order: bundled python libs (unittest...), external libs (werkzeug...), odoo API (models...), odoo modules (payment_paypal...).

File: addons/payment_paypal/tests/test_paypal.py [hunk 17]
```diff
@@ -47,18 +53,23 @@ def _get_expected_values(self):
...
     def test_redirect_form_values(self):
         tx = self._create_transaction(flow='redirect')
-        with mute_logger('odoo.addons.payment.models.payment_transaction'):
-            processing_values = tx._get_processing_values()
+        with patch(
+            "odoo.addons.payment.utils.generate_access_token",
+            new=self._generate_test_access_token,
+        ):
+            with mute_logger('odoo.addons.payment.models.payment_transaction'):
+                processing_values = tx._get_processing_values()
```
Comment (AntoineVDV) [inline addons/payment_paypal/tests/test_paypal.py]: These two context managers don't need to be one inside another; they can be chained like this: `with patch(...), patch(...):

File: addons/payment_paypal/tests/test_paypal.py [hunk 18]
```diff
@@ -47,18 +53,23 @@ def _get_expected_values(self):
...
 
         form_info = self._extract_values_from_html_form(processing_values['redirect_form_html'])
         self.assertEqual(
             form_info['action'],
             'https://www.sandbox.paypal.com/cgi-bin/webscr')
 
         expected_values = self._get_expected_values()
+        self.maxDiff = 1000
```
Comment (AntoineVDV) [inline addons/payment_paypal/tests/test_paypal.py:69]: Why is this required now?

File: addons/payment_paypal/tests/test_paypal.py [hunk 19]
```diff
@@ -71,11 +82,16 @@ def test_redirect_form_with_fees(self):
...
+        with patch(
+            "odoo.addons.payment.utils.generate_access_token",
+            new=self._generate_test_access_token,
+        ):
+            with mute_logger('odoo.addons.payment.models.payment_transaction'):
+                processing_values = tx._get_processing_values()
+            form_info = self._extract_values_from_html_form(processing_values['redirect_form_html'])
 
         self.assertEqual(form_info['action'], 'https://www.sandbox.paypal.com/cgi-bin/webscr')
+        self.maxDiff = None
```
Comment (AntoineVDV) [inline addons/payment_paypal/tests/test_paypal.py:95]: Why is this required now?

File: addons/payment_paypal/views/payment_provider_views.xml [hunk 20]
```diff
@@ -10,13 +10,8 @@
...
                            attrs="{'required':[('code', '=', 'paypal'), ('state', '!=', 'disabled')]}"/>
-                    <field name="paypal_seller_account" password="True"/>
-                    <!-- This field should no longer be used but is kept in debug mode for the time
-                         being, until we are sure that the verification protocol of IPN can be used
-                         for DPT notifications -->
-                    <field name="paypal_pdt_token" password="True"/>
-                    <field name="paypal_use_ipn"
-                           attrs="{'required':[('code', '=', 'paypal'), ('state', '!=', 'disabled')]}"/>
+                    <field name="paypal_pdt_token"  password="True"
+                           attrs="{'required': [('code', '=', 'paypal'), ('state', '!=', 'disabled')]}"/>
```
Comment (AntoineVDV) [inline addons/payment_paypal/views/payment_provider_views.xml:14]: Now that the PDT is required, we should be able to drop the `if not provider_sudo.paypal_pdt_token` check in the controller for the return from checkout.

**Review by AntoineVDV — CHANGES_REQUESTED** [review]

File: addons/payment_paypal/controllers/main.py [hunk 21]
```diff
@@ -32,9 +35,8 @@ def paypal_return_from_checkout(self, **pdt_data):
         the account. See https://developer.paypal.com/api/nvp-soap/payment-data-transfer/.
 
         The route accepts both GET and POST requests because PayPal seems to switch between the two
-        depending on whether PDT is enabled, whether the customer pays anonymously (without logging
```
[the remark this answers is not in the document]
Comment (AntoineVDV) [reply addons/payment_paypal/controllers/main.py:35]: The key is required but the admin might have not enabled the feature from PayPal side.

File: addons/payment_paypal/tests/test_paypal.py (hunk 18 above)
Comment (anko-odoo) [reply addons/payment_paypal/tests/test_paypal.py:69]: If I recall corectly it looks like something I forgot to remove.

**Review by AntoineVDV — CHANGES_REQUESTED** [review]

**Review by AntoineVDV — APPROVED** [review]
@robodoo override=ci/security
Reason: De-indent of safe existing code.
Link to ci/security build (master): https://runbot.odoo.com/runbot/build/24009600

@robodoo r+
